vcd/resource_vcd_independent_disk.go

```diff
@@ -128,25 +198,76 @@ func resourceVcdIndependentDiskCreate(d *schema.ResourceData, meta interface{})
 		return fmt.Errorf("error waiting to finish creation of independent disk: %s", err)
 	}
 
-	d.SetId(diskName)
+	diskHref := task.Task.Owner.HREF
+	disk, err := vdc.GetDiskByHref(diskHref)
+	if govcd.IsNotFound(err) {
+		log.Printf("unable to find disk with href %s: %s. Removing from state", diskHref, err)
+		d.SetId("")
+		return nil
+	}
+	if err != nil {
+		return fmt.Errorf("unable to find disk with href %s: %s", diskHref, err)
+	}
+
+	d.SetId(disk.Disk.Id)
 
 	return resourceVcdIndependentDiskRead(d, meta)
 }
 
 func resourceVcdIndependentDiskRead(d *schema.ResourceData, meta interface{}) error {
 	vcdClient := meta.(*VCDClient)
 
-	_, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+	_, vdc, err := vcdClient.GetOrgAndVdc("", d.Get("vdc").(string))
 	if err != nil {
 		return fmt.Errorf(errorRetrievingOrgAndVdc, err)
 	}
 
-	_, err = vdc.QueryDisk(d.Get("name").(string))
+	identifier := d.Id()
+	var disk *govcd.Disk
+	if identifier != "" {
+		disk, err = vdc.GetDiskById(identifier, true)
+		if govcd.IsNotFound(err) {
+			log.Printf("unable to find disk with ID %s: %s. Removing from state", identifier, err)
+			d.SetId("")
+			return nil
+		}
+		if err != nil {
+			return fmt.Errorf("unable to find disk with ID %s: %s", identifier, err)
+		}
+	} else {
+		identifier = d.Get("name").(string)
+		disks, err := vdc.GetDisksByName(identifier, true)
+		if govcd.IsNotFound(err) {
+			log.Printf("unable to find disk with ID %s: %s. Removing from state", identifier, err)
+			d.SetId("")
+			return nil
+		}
+		if err != nil {
+			return fmt.Errorf("unable to find disk with ID %s: %s", identifier, err)
+		}
+		if len(*disks) > 1 {
+			return fmt.Errorf("found more than one disk with ID %s: %s", identifier, err)
+		}
+		disk = &(*disks)[0]
```

Comment:
Is there a specific reason to dereference and take pointer of the value
Wouldn't it be the same as:
```
disk = disks[0]
```

Reply:
disk array is ref and element is ref too.
![image](https://user-images.githubusercontent.com/41415459/67272877-04a2f080-f4c6-11e9-9f44-1f9fa250e079.png)

Reply:
I misread it.